editor/src/components/canvas/controls/new-canvas-controls.tsx

```diff
@@ -269,15 +309,173 @@ const NewCanvasControlsInner = (props: NewCanvasControlsInnerProps) => {
     'NewCanvasControlsInner',
   )
 
-  const { localSelectedViews, localHighlightedViews, setLocalSelectedViews } = props
+  const {
+    localSelectedViews,
+    localHighlightedViews,
+    setLocalSelectedViews,
+    setLocalHighlightedViews,
+    setSelectionAreaRectangle,
+  } = props
   const cmdKeyPressed = keysPressed['cmd'] ?? false
 
   const contextMenuEnabled = !isLiveMode(editorMode)
   const { maybeHighlightOnHover, maybeClearHighlightsOnHoverEnd } = useMaybeHighlightElement()
 
-  const { onMouseMove, onMouseDown, onMouseUp } = useSelectAndHover(
-    cmdKeyPressed,
-    setLocalSelectedViews,
+  const selectModeHooks = useSelectAndHover(cmdKeyPressed, setLocalSelectedViews)
+
+  const ref = React.useRef<HTMLDivElement | null>(null)
+
+  const [mousePoint, setMousePoint] = React.useState<CanvasPoint | null>(null)
+  const [selectionAreaStart, setSelectionAreaStart] = React.useState<CanvasPoint | null>(null)
+
+  const storeRef = useRefEditorState((store) => {
+    return { jsxMetadata: store.editor.jsxMetadata }
+  })
+
+  const getCanvasPoint = React.useCallback(
+    (x: number, y: number): CanvasPoint => {
+      return windowToCanvasCoordinates(scale, canvasOffset, windowPoint({ x, y }))
+        .canvasPositionRounded
+    },
+    [scale, canvasOffset],
+  )
+
+  const mousePointOnCanvas = React.useMemo(() => {
+    if (mousePoint == null) {
+      return null
+    }
+    return getCanvasPoint(mousePoint.x, mousePoint.y)
+  }, [mousePoint, getCanvasPoint])
+
+  const selectionArea = React.useMemo((): CanvasRectangle | null => {
+    if (selectionAreaStart == null || mousePoint == null) {
+      return null
+    }
+    return makeSelectionArea(selectionAreaStart, mousePoint)
+  }, [mousePoint, selectionAreaStart])
+
+  const canSelectArea = React.useMemo((): boolean => {
+    if (selectionArea != null || isSelectModeWithArea(editorMode)) {
+      return true
+    }
+    if (editorMode.type !== 'select') {
+      return false
+    }
+    return true
```

Comment:
This can all be simplified to `return selectionArea != null || isSelectModeWithArea(editorMode)`, and shouldn't be a `useMemo` (it could either be inlined or become `function canSelectArea(...)`)

Reply:
yeah indeed, moved to a helper https://github.com/concrete-utopia/utopia/pull/3750/commits/b342869dcca26246b84df8bb7caca117f9082fcc